### PR TITLE
Recursive validation

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we enable the Dart validation code to check field messages with the `(validate) = true` option.

In order to obtain child message violations and store them in a variable in generated code, we introduce `LazyPreparation`, a new component to a validation `Rule`. This new codegen step generates code which goes before the constraint check. 

Example:

```proto
message DartCompiler {

    // ...

    Version version = 42 [(validate) = true];
}
```

Here is the (simplified) generated code for validating `DartCompiler.version`:

```dart
import 'package:spine_client/validate.dart' as v;

// ...

var violationsOf_version = v.validate(msg.version); // This line is generated by `LazyPreparation`.
if (violationsOf_version.isPresent) { // This line is generated by `LazyCondition`.
    violations.add(newViolation('Error message ...', ..., violationsOf_version.value)); // This line is generated by `LazyViolation`.
}
```

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: codegen/lib/src/constraint_violation.dart [hunk 1]
```diff
@@ -54,25 +57,39 @@ createViolationFactory(String standardPackage) {
...
+            ..named = true
             ..type = anyType
-            ..name = actualValue);
+            ..name = actualValueArg
+        );
+        var listOfViolations = TypeReference((b) => b
+            ..symbol = 'List'
+            ..types.add(violationTypeRef(standardPackage))
+        );
+        var childConstrainsParam = Parameter((b) => b
```
Comment (armiol) [inline codegen/lib/src/constraint_violation.dart]: `constraints`?

File: codegen/lib/src/field_validator_factory.dart [hunk 2]
```diff
@@ -304,3 +317,7 @@ typedef Expression LazyViolation(Expression fieldValue);
 ///  - `false` if the constraint obeyed.
 ///
 typedef Expression LazyCondition(Expression fieldValue);
+
+/// A function of a field value expresion to code which prepares context before the constraint
```
Comment (armiol) [inline codegen/lib/src/field_validator_factory.dart]: Please move the phrase starting with `before` to the next line. This won't leave a "check" alone in the dark.

Comment (armiol) [inline codegen/lib/src/field_validator_factory.dart]: `expression`?

And honestly, I don't understand.
Can you please explain in more detail and maybe have an example?

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov please see my comments. I am not able to tell I understand the idea completely. But it's better than I thought it would be — given my lack of Dart experience :)

Please consider putting more details into the PR description telling of the under-the-hood actions. For instance, you could tell a small story and incorporate that compilation restriction we have, etc.

File: client/src/test/proto/spine/test/tools/dart/validation.proto [hunk 3]
```diff
@@ -79,3 +79,17 @@ message LotteryTicket {
...
+    string value = 1 [(required) = true, (pattern).regex = "\\w{20}"];
+}
+
+message WorkInProgressSnapshot {
+
+    map<string, TaskId> assignment = 1 [(validate) = true];
+
+    LocalTime when = 2 [(validate) = true];
+
+    repeated TaskId backlog = 3 [(validate) = true];
```
Comment (alexander-yevsyukov) [inline client/src/test/proto/spine/test/tools/dart/validation.proto:102]: Please document this field. Its purpose isn't quite clear.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM

**Review by alexander-yevsyukov — APPROVED** [review]
LGTM

**Pull request merged**